cuDF error handling document (#7917)
rapidsai/cudf: docs/cudf/source/errors.md

# cuDF Error Handling Documentation (draft)

## Scope

This document serves to guide the use of error types when dealing with different exceptions encountered within cuDF domain. As of the draft version, it also serves as a reference for supporting libraries (like libcudf) to boost awareness of common error types encountered and to standardize their error interface. Note that this document will not serve as a complete list of error types thrown by cuDF.

## Overview

> On a high level, exceptions consist of exception type and exception payload. Each can be standardized by the API. cuDF only standardizes exception type and uses the payload to convey debugging information for users. (Subject to discussion: is it sufficient to only use builtin error types but not the payload to signal all errors?)

Reviewer: A related question: since we aim to match pandas API, do we consider it sufficient to match API up to the exception type alone, or do we also need to match the message? I think type is sufficient, but we should standardize around that.

Reviewer: I'm not sure I understand exactly what your discussion question is asking. Are you asking whether the payload should be promised by the API as well?

Author: @vyasr yes that's the question: should the API also promise the error payload? Take `ValueError` for example, the semantic could vary even for the same API: e.g.
https://github.com/rapidsai/cudf/blob/4893259b2fd6cf1f2079eff68249290708519892/python/cudf/cudf/core/column/numerical.py#L564
https://github.com/rapidsai/cudf/blob/4893259b2fd6cf1f2079eff68249290708519892/python/cudf/cudf/core/column/numerical.py#L580

Reviewer: I think promising a consistent payload is a nice-to-have, but not a must. In an ideal world, yes it should be consistent, but given the current maturity of the project and the fast evolution it is probably too much to ask. I'd say we could reassess this if/when the data APIs are mature enough that we can decouple somewhat from pandas, but for now our decision should be governed by whatever pandas promises. @shwina was fine with having our exception deviate from pandas in some recent PR of mine that he approved, so he may also have thoughts here.

|     ValueError     | The value of the object does not satisfy operation requirement. Usually raising this error requires data introspection. In cases where the object only accept from a finite set of values (such as enum or string set), also use `ValueError`. | [The dataframe to describe is empty](https://github.com/rapidsai/cudf/blob/7d49f75df9681dbe1653029e7d508355884a6d86/python/cudf/cudf/core/dataframe.py#L5366) |
|     IndexError     | Array access out of range | [Retrieving rows from a column specified by an out of bound index](https://github.com/rapidsai/cudf/blob/7d49f75df9681dbe1653029e7d508355884a6d86/python/cudf/cudf/core/column/column.py#L849-L851) |
|     KeyError       | Mapping access with invalid key | [Retrieving rows from column specified by an invalid key](https://github.com/rapidsai/cudf/blob/7d49f75df9681dbe1653029e7d508355884a6d86/python/cudf/cudf/core/indexing.py#L177) |
| NotImplementedError| Operation of object is planned, but not yet supported | (none) |
|     IOError        | (TBA by cuIO) | (TBA) |

> 

Reviewer: Other less important errors to consider:
- `OverflowError` for numerical overflow. I believe that we recently discussed just silently allowing overflow, with the possibility of moving long-term towards the (IMO better) behavior of numpy which will at least warn (but not error) on overflow.
- Other I/O related errors (`FileNotFoundError`, `OSError`, etc). If we're rolling our own writers (or using other C++ libraries) there may be cases where we need to map the error more precisely than just `IOError`.

Custom error types, should not be used whenever possible.

> Error raised by the API is considered as part of the API interface and should be documented in the docstring. For a specific invalid input, if the error raised has changed, it is considered as a `breaking` change.

Reviewer: ```suggestion
Errors raised by the API are considered part of the API and should be documented in the docstring. For a specific invalid input, changing the raised exception is considered a breaking change.
```

## Handling Exceptions Thrown by Supporting Libraries

cuDF depends on external libraries to function. While cuDF does extensive tests on inputs to make sure data passed to lower level is legal, supporting libraries may also perform their own data checks. Very often such checks are redundant. Double-checking adds unecessary latency to cuDF API calls and is thus unwanted. To allow reusing error checks from supporting libraries, the following section explains how cuDF proposes to map errors from supporting libraries to cuDF errors.

Note that this does not mean cuDF will skip checking errors. As supporting libraries have different performance constraints to cuDF, cuDF should perform checks when necessary. An example is that libcudf [does not introspect data](https://github.com/rapidsai/cudf/issues/5505), but cuDF should.

### Python Libraries

> If python libraries throws a builtin error type, cuDF will surface such error.

Reviewer: ```suggestion
If Python libraries raise a builtin error type, cuDF will surface that error.
```

> If python libraries throws a custom error type, cuDF will attempt to reinterpret it into builtin error type.

Reviewer: We may want to establish firmer rules around what this means and how we do it. I suspect that there will be some cases where this makes sense, while in others we would want to just forward the exception along. Do you have particular exceptions in mind for this? I think we'd want to see examples to decide the appropriate behavior here.

Author: Yeah, it was pyarrow's error types that led me to come up with this section: https://github.com/apache/arrow/blob/master/python/pyarrow/error.pxi

Reviewer: Personally I don't think we need to codify this. There may be cases where Python libraries throw builtin types, but we know the reason why and might catch and either handle or reinterpret as a different, more specific and informative error. Similarly, there may be cases when a custom error type thrown by an underlying Python library could actually signal a broader class of problems in our context, so we just pass it through transparently. As long as we are consistent with a particular API over time, I think we should give ourselves the flexibility to handle errors from underlying libraries on an as-needed basis.

### C++ Libraries

> cuDF interfaces with c++ libraries through Cython. By default, cuDF builds such interfaces capturing all exceptions thrown by c++ libraries. Cython maps these exceptions into python errors as outlined by [this document](http://docs.cython.org/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions).

Reviewer: ```suggestion
cuDF interfaces with C++ libraries through Cython. By default, cuDF captures all exceptions thrown by C++ libraries. Cython maps these exceptions into Python errors as outlined by [this document](http://docs.cython.org/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions).
```

> In an unambiguous situation, for example, the c++ function called by cuDF will only throw one instance of `std::out_of_range` exception, indicating an out of bound array access error, cuDF will skip checks in the python level and surface such error (through automatic Cython error mapping) if occurs.
> 
> In an ambiguous situation, for example, the c++ function may raise two instances of `std::invalid_argument` error, each signaling different kinds of invalid argument combination, cuDF will check the invalid combinations and re-raise with proper user message.

Reviewer: ```suggestion
In cases where the meaning of a C++ exception can be unambiguously determined, cuDF will skip any checks in Python and directly surface such exceptions via Cython's C++->Python exception mapping. However, in some cases the exceptions may be ambiguous. For example, a C++ function may raise two instances of `std::invalid_argument` error, each signaling different kinds of invalid argument combination. In that case, cuDF will check the invalid combinations and re-raise with proper user message.
```

> In cases when the contents of `what()` is standardized by the supporting library, cuDF will utilize it to disambiguate.

Reviewer: ```suggestion
In cases when the contents of [`what()`](https://www.cplusplus.com/reference/exception/exception/what/) is standardized by the supporting library, cuDF will utilize it to disambiguate.
```
